// zos/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TestHelper = exports.runWithTruffle = exports.stdout = exports.naming = exports.scripts = exports.commands = exports.network = exports.local = exports.files = exports.version = undefined;

var _commands = require('./commands');

var _commands2 = _interopRequireDefault(_commands);

var _scripts = require('./scripts');

var _scripts2 = _interopRequireDefault(_scripts);

var _files = require('./models/files');

var _files2 = _interopRequireDefault(_files);

var _local = require('./models/local');

var _local2 = _interopRequireDefault(_local);

var _network = require('./models/network');

var _network2 = _interopRequireDefault(_network);

var _TestHelper = require('./models/TestHelper');

var _TestHelper2 = _interopRequireDefault(_TestHelper);

var _naming = require('./utils/naming');

var _naming2 = _interopRequireDefault(_naming);

var _runWithTruffle = require('./utils/runWithTruffle');

var _runWithTruffle2 = _interopRequireDefault(_runWithTruffle);

var _stdout = require('./utils/stdout');

var _stdout2 = _interopRequireDefault(_stdout);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// module information
const version = 'v' + require('../package.json').version;

// commands


// model objects


// utils

const stdout = { log: _stdout2.default, silent: _stdout.silent };

exports.version = version;
exports.files = _files2.default;
exports.local = _local2.default;
exports.network = _network2.default;
exports.commands = _commands2.default;
exports.scripts = _scripts2.default;
exports.naming = _naming2.default;
exports.stdout = stdout;
exports.runWithTruffle = _runWithTruffle2.default;
exports.TestHelper = _TestHelper2.default;
