// zos/lib/commands/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _add = require('./add');

var _add2 = _interopRequireDefault(_add);

var _bump = require('./bump');

var _bump2 = _interopRequireDefault(_bump);

var _check = require('./check');

var _check2 = _interopRequireDefault(_check);

var _create = require('./create');

var _create2 = _interopRequireDefault(_create);

var _freeze = require('./freeze');

var _freeze2 = _interopRequireDefault(_freeze);

var _init = require('./init');

var _init2 = _interopRequireDefault(_init);

var _link = require('./link');

var _link2 = _interopRequireDefault(_link);

var _publish = require('./publish');

var _publish2 = _interopRequireDefault(_publish);

var _push = require('./push');

var _push2 = _interopRequireDefault(_push);

var _remove = require('./remove');

var _remove2 = _interopRequireDefault(_remove);

var _session = require('./session');

var _session2 = _interopRequireDefault(_session);

var _setAdmin = require('./set-admin');

var _setAdmin2 = _interopRequireDefault(_setAdmin);

var _status = require('./status');

var _status2 = _interopRequireDefault(_status);

var _unlink = require('./unlink');

var _unlink2 = _interopRequireDefault(_unlink);

var _update = require('./update');

var _update2 = _interopRequireDefault(_update);

var _verify = require('./verify');

var _verify2 = _interopRequireDefault(_verify);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
  add: _add2.default,
  bump: _bump2.default,
  check: _check2.default,
  create: _create2.default,
  freeze: _freeze2.default,
  init: _init2.default,
  link: _link2.default,
  publish: _publish2.default,
  push: _push2.default,
  remove: _remove2.default,
  session: _session2.default,
  setAdmin: _setAdmin2.default,
  status: _status2.default,
  unlink: _unlink2.default,
  update: _update2.default,
  verify: _verify2.default
};
